pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -346,6 +350,17 @@ func NewMaxPDVolumeCountPredicate(filterName string, pvInfo PersistentVolumeInfo
 	return c.predicate
 }
 
+// NewMaxAttachableVolumeLimitPredicate returns a predicate that can count maximum number of attachable volumes
+func NewMaxAttachableVolumeLimitPredicate(
```

Comment:
This feature being Alpha we can't perhaps remove existing predicates, but we can document how user can use just this predicate via scheduler policy file that lists this predicate but drops others. Alternately when alpha feature gate is enabled and volume limits are available from node objects, we can make other predicates NOOP.

Reply:
I think when this feature is enabled you could do something like this: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/defaults/defaults.go#L177-#L180 to disable those old predicates.

Reply:
I think we may not want to remove the old predicates if feature gate is enabled because it is possible that schduler is running on 1.11 and with feature gate but nodes do not have feature gate or is running with older version.

I will make other predicates instead NOOP if feature gate is enabled and node's allocatable property contains volume limits.

Reply:
Also wondering, why not just use the existing predicates instead of creating a new one? My understanding after discussion with you was that: 1) volumelimits are exposed via kubelet in allocatable that would be used by the scheduler . 2) so for these existing predicate instead of using current hardcoded limits, the dynamically limits set in step 1 would be used.

Reply:
There are three existing predicates `MaxEBSVolumeCountPred` , `MaxGCEPDVolumeCountPred` and `MaxAzureDiskVolumeCountPred` which work only for those 3 volume types (EBS, GCE-PD and Azure).

This PR presents a design which will make VolumeAttachableLimit predicate work for ANY volume type, as long as they implement necessary interface.  

If we updated existing predicate, it will still only only to those 3 volume types and won't work for new volume types like CSI or anything else.

Reply:
> My understanding after discussion with you was that: 1) volumelimits are exposed via kubelet in allocatable that would be used by the scheduler . 2) so for these existing predicate instead of using current hardcoded limits, the dynamically limits set in step 1 would be used.

if the scheduler was making use of these resources in the same way as existing resources that would work. the reason that doesn't work here is that nothing in the API indicates that resource is consumed... the scheduler has to infer it from the volume types

Reply:
To keep the integration between volumeManager and scheduler low, would this flow work?: 1) have an admission plugin  that adds (integer) resources request to an input pod based on volume types in the pod spec by calling volumeManager 2) scheduler just works based on those added resources requests and dynamically added limits via kubelet. Overall It would also help with avoiding any overhead (if any) related to calling volumeManager.  Just throwing it as an idea.

Reply:
Yeah that was discussed and we chose not to implement that because:

a. We will have to make an API change to pod spec to store such a counted resource somewhere. It was discussed that, if we made such a API change - it will be probably only used for counting volumes and hence it is bit premature to add such a field to pod spec. If there were other resources that needed to be counted at pod level, we can consider the API change.  This proposal however does not excludes taking this route if we decide in future. in Beta for example - we can certainly do this, if that is path we want to take.
b. There was already a precedence for using resources that are available in allocatable but not explicitly requestable by pods (max pod count). sorry @liggitt I forgot about `pod` field in allocatable, when we discussed this yesterday. :( 

Also - I think we can remove integration between volume plugin manager and scheduler, I responded this https://github.com/kubernetes/kubernetes/pull/64154#discussion_r190955132 . I will update the proposal.

Reply:
> We will have to make an API change to pod spec to store such a counted resource somewhere. It was discussed that, if we made such a API change - it will be probably only used for counting volumes and hence it is bit premature to add such a field to pod spec.

pinned resource sets for the in-tree volume use seems ok. I'm much more interested in the representation of resource consumption (including attach count consumption) for CSI volumes, which are only in PV objects, not in the pod spec.

Reply:
Fixed the original comment. This PR no longer introduces a new predicate. Because this PR only handles GCE, AWS and Azure and there are already predictes for that, this PR merely changes source of max Volume count.